dm/worker/source_worker.go

```diff
@@ -186,11 +207,54 @@ func (w *Worker) Close() {
 	}
 
 	w.closed.Store(true)
+
+	w.sourceDB.Close()
+	w.sourceDB = nil
+
 	w.l.Info("Stop worker")
 }
 
+// updateSourceStatus updates w.sourceStatus.
+func (w *SourceWorker) updateSourceStatus(ctx context.Context) error {
+	w.sourceDBMu.Lock()
```

Comment:
This lock doesn't actually work because we directly use `w.sourceDB.DB` in the following part.

Reply:
sql.DB is safe for concurrent use, I only want to protect the assignment to `w.sourceDB`

Reply:
If one routine assign value to w.sourceDB in L219 and another read `w.sourceDB` at L233, we still get a data race.

Reply:
Besides, we will unlock `w.sourceDBMu` twice if we successfully apply a new DB.

Reply:
if the goroutine is going to read `w.sourceDB` at L244 (updated line numbers 🤣) it must pass the nil check in 231 or itself initiated `w.sourceDB`. For the former case, since it has been passed nil check, the assignment is already done. For the latter case, this goroutine finished the assignment and release the lock, so other goroutine which acquired the lock will pass the nil check and not go to line 233 again

Reply:
I see. What if we `QueryStatus` and `worker.Close` at the same time? Will L244 cause a data race with L223?

Reply:
currently we can't assure that, will update later

Reply:
I found `SourceWorker.RWMutex` has protected this.

Reply:
Will this cause a problem?
```
QueryStatus ───► closed(false) ───► utils.GetMasterStatus

worker.Close ─────────────────────► w.sourceDB = nil
```

Reply:
Close will call w.Lock() and QueryStatus will call w.RLock()